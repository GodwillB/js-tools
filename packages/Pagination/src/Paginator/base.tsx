/** A bootstrap powered pagination component
 * inspired by https://scotch.io/tutorials/build-custom-pagination-with-react
 */
import React, { Fragment, MouseEvent, useState } from 'react';
import { Pagination, PaginationItem, PaginationLink } from 'reactstrap';
import { range } from './utils';

/** interface for argument passed to the opPageChange prop */
export interface PaginationData {
  currentPage: number;
  pageLimit: number;
  totalPages: number;
  totalRecords: number;
}

/** interface for Paginator props */
export interface PaginatorProps {
  totalRecords: number;
  pageLimit: number;
  pageNeighbours: number;
  onPageChange: (e: PaginationData) => void;
  ariaLabel?: string;
}

/** default props for paginator component */
const defaultPaginatorProps: PaginatorProps = {
  ariaLabel: 'page navigation',
  onPageChange: f => f,
  pageLimit: 30,
  pageNeighbours: 2,
  totalRecords: 0
};

/** the pagination component */
const Paginator = (props: PaginatorProps) => {
  const { totalRecords, pageLimit, pageNeighbours, ariaLabel, onPageChange } = props;

  let totalPages = 0;

  const [currentPage, setCurrentPage] = useState<number>(1);
  // pageNeighbours can be in [0, 1, 2, 3, 4, 5]
  const neighbourPillsNum = Math.max(0, Math.min(pageNeighbours, 5));

  totalPages = Math.ceil(totalRecords / pageLimit);

  const pages = fetchPageNumbers(neighbourPillsNum, totalPages, currentPage);
  // click handlers

  /** click Handler; registers the number of the clicked pagination Item
   * handles click event
   * @param {number} - number of the clicked plaginationItem
   * @return {(e: MouseEvent) => void }
   */
  const handleClick = (page: number) => (e: MouseEvent) => {
    e.preventDefault();
    gotoPage(page);
  };

  /** click handler for when user clicks on previous
   * @param {MouseEvent} e - MouseEvent
   */
  const handleMoveLeft = (e: MouseEvent) => {
    e.preventDefault();
    gotoPage(currentPage - 1);
  };

  /** click handler for when user clicks on next
   * @param {MouseEvent} e - MouseEvent
   */
  const handleMoveRight = (e: MouseEvent) => {
    e.preventDefault();
    gotoPage(currentPage + 1);
  };

  /** called when user click on a certain pagination item
   * calls the onPageChange handler
   * @param {number} page -  the number of the page clicked
   */
  const gotoPage = (page: number) => {
    const thisPage = Math.max(1, Math.min(page, totalPages));
    const paginationData = {
      currentPage: thisPage,
      pageLimit,
      totalPages,
      totalRecords
    };
    onPageChange(paginationData);
    setCurrentPage(thisPage);
  };

  return (
    <Fragment>
      <Pagination aria-label={ariaLabel} size="sm">
        <PaginationItem className={`page-item ${currentPage > 1 ? '' : 'disabled'}`}>
          <PaginationLink
            className={`page-link`}
            href="#"
            aria-label="Previous"
            onClick={handleClick(1)}
          >
            <span aria-hidden="true">start</span>
            <span className="sr-only">Start</span>
          </PaginationLink>
        </PaginationItem>
        <PaginationItem className={`page-item ${currentPage > 1 ? '' : 'disabled'}`}>
          <PaginationLink
            className={`page-link`}
            href="#"
            aria-label="Previous"
            onClick={handleMoveLeft}
          >
            <span aria-hidden="true">previous</span>
            <span className="sr-only">Previous</span>
          </PaginationLink>
        </PaginationItem>
        {pages.map((page, index) => {
          return (
            <PaginationItem
              key={index}
              className={`page-item ${currentPage === page ? ' active' : ''}`}
            >
              <PaginationLink className="page-link" href="#" onClick={handleClick(page as number)}>
                {page}
              </PaginationLink>
            </PaginationItem>
          );
        })}

        <PaginationItem className={`page-item  ${currentPage < totalPages ? '' : 'disabled'}`}>
          <PaginationLink
            className={`page-link`}
            href="#"
            aria-label="Next"
            onClick={handleMoveRight}
          >
            <span aria-hidden="true">next</span>
            <span className="sr-only">Next</span>
          </PaginationLink>
        </PaginationItem>
        <PaginationItem className={`page-item ${currentPage < totalPages ? '' : 'disabled'}`}>
          <PaginationLink
            className={`page-link ${currentPage < totalPages ? '' : 'disabled'}`}
            href="#"
            aria-label="Previous"
            onClick={handleClick(totalPages)}
          >
            <span aria-hidden="true">end</span>
            <span className="sr-only">End</span>
          </PaginationLink>
        </PaginationItem>
      </Pagination>
    </Fragment>
  );
};

// helper && utilities

/** Let's say we have 10 pages and we set neighbourPillsNum to 2
 * Given that the current page is 6
 * The pagination control will look like the following:
 *
 * (start) previous {4 5} [6] {7 8} next (end)
 *
 * (x) => terminal pages: first and last page(always visible)
 * [x] => represents current page
 * {...x} => represents page neighbours
 *
 * @param {number} neighbourPillsNum - the max number of links/pills on either side of current page
 * @param {number} totalPages - the total number of pages
 * @param {number} currentPage - the currently selected page
 * @return {number []} - Array of numbers that will form the pills between previous and next
 */
export const fetchPageNumbers = (
  neighbourPillsNum: number,
  totalPages: number,
  currentPage: number
) => {
  const startPage = Math.max(1, currentPage - neighbourPillsNum);
  const endPage = Math.min(totalPages + 1, currentPage + neighbourPillsNum + 1);

  const numberedPages: number[] = range(startPage, endPage);

  return numberedPages;
};

Paginator.defaultProps = defaultPaginatorProps;

export { Paginator };
