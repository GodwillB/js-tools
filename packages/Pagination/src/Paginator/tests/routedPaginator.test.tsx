import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import { BasePaginator } from '../base';
import { RoutedPaginator } from '../routedPaginator';

const history = createBrowserHistory();

describe('packages/paginator/routedPaginator', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(
      <Router history={history}>
        <RoutedPaginator />
      </Router>
    );
  });

  it('renders correctly', () => {
    const props = {
      history,
      match: {
        isExact: true,
        params: {
          tablePage: 2
        },
        path: '/records/:tablePage',
        url: '/records/2'
      },
      pageLimit: 30,
      totalRecords: 135
    };

    const wrapper = mount(
      <Router history={history}>
        <RoutedPaginator {...props} />
      </Router>
    );
    const basePaginatorWrapper = wrapper.find(BasePaginator);
    expect(toJson(basePaginatorWrapper)).toMatchSnapshot('base paginator');
  });

  it('passes corect page from url case 1', () => {
    const mock: any = jest.fn();
    const props = {
      history,
      location: mock,
      match: {
        isExact: true,
        params: {
          tablePage: 2
        },
        path: '/records/:tablePage',
        url: '/records/2'
      },
      pageLimit: 40,
      totalRecords: 135
    };
    const wrapper = mount(
      <Router history={history}>
        <RoutedPaginator {...props} />
      </Router>
    );
    const basePaginatorWrapper = wrapper.find(BasePaginator);
    expect(basePaginatorWrapper.props().currentPage).toEqual(2);
  });
});
