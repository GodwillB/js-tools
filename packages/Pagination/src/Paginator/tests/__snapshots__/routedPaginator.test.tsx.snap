// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages/paginator/routedPaginator renders correctly: base paginator 1`] = `
<BasePaginator
  ariaLabel="page navigation"
  currentPage={1}
  endLabel="End"
  nextLabel="Next"
  pageChangeHandler={[Function]}
  pageLimit={30}
  pageNeighbours={2}
  previousLabel="Previous"
  startLabel="Start"
  totalRecords={135}
>
  <Pagination
    aria-label="page navigation"
    listTag="ul"
    size="sm"
    tag="nav"
  >
    <nav
      aria-label="page navigation"
      className=""
    >
      <ul
        className="pagination pagination-sm"
      >
        <PaginationItem
          className="page-item disabled"
          tag="li"
        >
          <li
            className="page-item disabled page-item"
          >
            <PaginationLink
              aria-label="Previous"
              className="page-link"
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                aria-label="Previous"
                className="page-link page-link"
                href="#"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  Start
                </span>
                <span
                  className="sr-only"
                >
                  Start
                </span>
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
        <PaginationItem
          className="page-item disabled"
          tag="li"
        >
          <li
            className="page-item disabled page-item"
          >
            <PaginationLink
              aria-label="Previous"
              className="page-link"
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                aria-label="Previous"
                className="page-link page-link"
                href="#"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  Previous
                </span>
                <span
                  className="sr-only"
                >
                  Previous
                </span>
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
        <PaginationItem
          className="page-item  active"
          key="0"
          tag="li"
        >
          <li
            className="page-item  active page-item"
          >
            <PaginationLink
              className="page-link"
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                className="page-link page-link"
                href="#"
                onClick={[Function]}
              >
                1
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
        <PaginationItem
          className="page-item "
          key="1"
          tag="li"
        >
          <li
            className="page-item  page-item"
          >
            <PaginationLink
              className="page-link"
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                className="page-link page-link"
                href="#"
                onClick={[Function]}
              >
                2
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
        <PaginationItem
          className="page-item "
          key="2"
          tag="li"
        >
          <li
            className="page-item  page-item"
          >
            <PaginationLink
              className="page-link"
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                className="page-link page-link"
                href="#"
                onClick={[Function]}
              >
                3
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
        <PaginationItem
          className="page-item  "
          tag="li"
        >
          <li
            className="page-item   page-item"
          >
            <PaginationLink
              aria-label="Next"
              className="page-link"
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                aria-label="Next"
                className="page-link page-link"
                href="#"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  Next
                </span>
                <span
                  className="sr-only"
                >
                  Next
                </span>
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
        <PaginationItem
          className="page-item "
          tag="li"
        >
          <li
            className="page-item  page-item"
          >
            <PaginationLink
              aria-label="Previous"
              className="page-link "
              href="#"
              onClick={[Function]}
              tag="a"
            >
              <a
                aria-label="Previous"
                className="page-link  page-link"
                href="#"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  End
                </span>
                <span
                  className="sr-only"
                >
                  End
                </span>
              </a>
            </PaginationLink>
          </li>
        </PaginationItem>
      </ul>
    </nav>
  </Pagination>
</BasePaginator>
`;
